lib/chef/resource/aws_route53_hosted_zone.rb

```diff
@@ -0,0 +1,220 @@
+
+      # AWS stores some attributes off to the side here.
+      hosted_zone_config = make_hosted_zone_config(new_resource)
+
+      values = {
+        name: new_resource.name,
+        hosted_zone_config: hosted_zone_config,
+        caller_reference: "chef-provisioning-aws-#{SecureRandom.uuid.upcase}",  # required, unique each call
+      }
+
+      zone = new_resource.driver.route53_client.create_hosted_zone(values).hosted_zone
+      new_resource.aws_route53_zone_id(zone.id)
+
+      begin
+        record_set_resources = get_record_sets_from_resource(new_resource)
+
+        if record_set_resources
+          change_list = record_set_resources.map { |rs| rs.to_aws_change_struct(CREATE) }
+
+          new_resource.driver.route53_client.change_resource_record_sets(hosted_zone_id: new_resource.aws_route53_zone_id,
+                                                                         change_batch: {
+                                                                           comment: RRS_COMMENT,
+                                                                           changes: change_list,
+                                                                           })
+        end
+      rescue => ex
+        # the change call is transactional, so we just need to clean up the zone we created.
+        new_resource.driver.route53_client.delete_hosted_zone(id: zone.id)
+        raise
+      end
+      zone
+    end
+  end
+
+  def update_aws_object(hosted_zone)
+    converge_by "update Route 53 zone #{new_resource}" do
+      new_resource.aws_route53_zone_id(hosted_zone.id)
+
+      if new_resource.comment != hosted_zone.config.comment
+        new_resource.driver.route53_client.update_hosted_zone_comment(id: hosted_zone.id, comment: new_resource.comment)
+      end
+
+      begin
+        record_set_resources = get_record_sets_from_resource(new_resource)
+        if record_set_resources
+          aws_record_sets = hosted_zone.resource_record_sets
+
+          change_list = []
+
+          # TODO: the SOA and NS records have identical :name properties (the zone name), so one of them will
+          # be overwritten in the `keyed_aws_objects` hash. mostly we're declining to operate on SOA and NS,
+          # so it probably doesn't matter, but bears investigating.
+
+          # we already checked for duplicate Chef RR resources in #get_record_sets_from_resource.
+          keyed_chef_resources = record_set_resources.reduce({}) { |coll, rs| (coll[rs.aws_key] ||= []) << rs; coll }
+          keyed_aws_objects    = aws_record_sets.reduce({})      { |coll, rs| coll[rs.aws_key] = rs; coll }
+
+          keyed_chef_resources.each do |key, chef_resource_ary|
+            chef_resource_ary.each do |chef_resource|
+              # RR already exists...
+              if keyed_aws_objects.has_key?(key)
+                # ... but we want to delete or update it.
+                action = chef_resource.action.first == :destroy ? DELETE : UPDATE
```

Comment:
Will this always perform an update even if the record set already matches the resource?  I don't see the logic where we diff the current state to the desired state to see if we need an update.